component/stream/stream_cache.go

```diff
@@ -200,3 +210,104 @@ func (c *cache) filePurge(key, value interface{}) {
 	c.blocks.Remove(key)
 	c.evictedBlock = blockKey{}
 }
+
+// Using key construct a file name for persisted block
+func (c *cache) getLocalFilePath(key blockKey) string {
+	return filepath.Join(c.diskPath, key.fileKey+"__"+fmt.Sprintf("%d", key.offset)+"__")
+}
+
+// Persist this block on disk
+func (c *cache) persistBlockOnDisk(block *cacheBlock, key blockKey) {
+	localPath := c.getLocalFilePath(key)
+
+	log.Debug("streamcache::persistBlockOnDisk : Saving file %s offset %d to disk", key.fileKey, key.offset)
+
+	err := os.MkdirAll(filepath.Dir(localPath), os.FileMode(0775))
+	if err != nil {
+		log.Err("streamcache::persistBlockOnDisk : unable to create local directory %s [%s]", localPath, err.Error())
+		return
+	}
+
+	f, err := os.OpenFile(localPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0775)
+	if err != nil {
+		log.Err("streamcache::persistBlockOnDisk : Failed to create file for persisting block on disk %s (%s)", localPath, err.Error())
+		return
+	}
+	f.Write(block.data)
+	f.Close()
+
+	c.diskBlocks.Set(key, true)
+}
+
+// Load data for this block from disk
+func (c *cache) getBlockFromDisk(block *cacheBlock, key blockKey) bool {
+	localPath := c.getLocalFilePath(key)
+	info, err := os.Stat(localPath)
+
+	if err != nil {
+		return false
+	}
+
+	log.Debug("streamcache::getBlockFromDisk : Reading block for %s offset %d from disk", key.fileKey, key.offset)
+	if time.Since(info.ModTime()).Seconds() > c.diskTimeoutSec {
+		// File exists on local disk but disk cache timeout has elapsed
+		os.Remove(localPath)
+		c.diskBlocks.Remove(key)
+		return false
+	}
+
+	f, err := os.OpenFile(localPath, os.O_RDONLY, 0775)
+	if err != nil {
+		return false
+	}
+
+	f.Read(block.data)
+	f.Close()
+	os.Remove(localPath)
+
+	c.diskBlocks.Remove(key)
+
+	return true
+}
+
+// Remove this block from in-memory cache
+func (c *cache) evictBlock(key blockKey, block *cacheBlock) {
+	// clean the block data to not leak any memory
+	block.Lock()
+
+	if c.persistance {
+		c.persistBlockOnDisk(block, key)
+	}
+
+	block.data = nil
+	c.evictedBlock = key
+	block.Unlock()
+}
+
+// Remove this block from disk cache
+func (c *cache) evictDiskBlock(key blockKey, val bool) {
+	// clean the block data to not leak any memory
+	localPath := c.getLocalFilePath(key)
+	err := os.Remove(localPath)
+	if err != nil {
+		log.Err("streamcache::evictDiskBlock : Failed to delete file for persisted block %s (%s)", localPath, err.Error())
+	}
+
+	dirPath := filepath.Dir(localPath)
+	if dirPath != c.diskPath {
+		os.Remove(filepath.Dir(localPath))
+	}
+}
+
+func (c *cache) wipeoutDiskCache() {
+	log.Trace("streamcache::evictDiskBlock : Wipe out disk cache in progress")
+
+	dirents, err := os.ReadDir(c.diskPath)
+	if err != nil {
+		return
+	}
+
+	for _, entry := range dirents {
+		os.RemoveAll(filepath.Join(c.diskPath, entry.Name()))
```

Comment:
Why not just do a os.RemoveAll of c.diskPath?

Reply:
We might not have created this directory so removing it will not make sense, so we are just iterating its children and then deleting all of them,